### ISSUE TITLE
Docker reporting 'bad certificate' with docker commands

### ISSUE THREAD

**Issue by ravishtiwari** [docker/toolbox#255]
Hi,
I am getting following error with any docker command that I run:
```
docker run hello-world
The server probably has client authentication (--tlsverify) enabled. Please check your TLS client certification settings: Post https://192.168.99.100:2376/v1.20/containers/create: remote error: bad certificate
```
 Even with simple **docker version** command. 
My system info: 
OS : **Mac OS X Yosemite (10.10.5)**
docker version: 
``` 
1.8.3
```
docker info: 
```
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 18:01:15 UTC 2015
 OS/Arch:      darwin/amd64
```
*uname -a:*
```
Darwin ravishk.local 14.5.0 Darwin Kernel Version 14.5.0: Tue Sep  1 21:23:09 PDT 2015; root:xnu-2782.50.1~1/RELEASE_X86_64 x86_64
```

*Steps to reproduce*
- Install docker package from docker.com, using following guide http://docs.docker.com/mac/step_one/
- Go to Terminal and do *docker run docker-centos-hello* OR *docker run hello-world*  OR any other container of your choice

You'll the TLS verify error, instead output stated in the getting started guide. I've used Docker on Ubuntu earlier, but, never faced this issue before.

Ravish

**Comment by ravishtiwari** [comment]
I reported this here as well https://github.com/docker/docker/issues/17314, but, I was advised to raise it in Toolbox instead.

**Comment by FrenchBen** [comment]
It looks like the certificate wasn't generated properly. I would try to do the following from a terminal to fix it:
`docker-machine -D regenerate-certs default`

Otherwise you can always remove the VM and let kitematic re-create it:
`docker-machine rm default`
Then open Kitematic.

**Comment by sf-wind** [comment]
I have the same issue. I tried the two commands and nothing works. I can't make Kitematic work also. It reports some other error... 

This is the first time I try docker, just followed the instructions. The experience is not so smooth...

**Comment by FrenchBen** [comment]
@sf-wind Did you install everything using the [Docker Toolbox](https://www.docker.com/toolbox) ?

**Comment by sf-wind** [comment]
Yes, I did.

**Comment by FrenchBen** [comment]
Can you show me the output of:
`docker-machine ls`
Also can you share the version of the Docker Toolbox that you are running?

**Comment by sf-wind** [comment]
The output is:
NAME      ACTIVE   DRIVER       STATE     URL                         SWARM
default   *        virtualbox   Running   tcp://192.168.99.100:2376   

The toolbox is : DockerToolbox-1.8.3.pkg

**Comment by FrenchBen** [comment]
That looks correct. 
Couple more:
`docker-machine env default`
`docker-machine ssh default docker ps -a`

**Comment by sf-wind** [comment]
The first one:

    export DOCKER_TLS_VERIFY="1"
    export DOCKER_HOST="tcp://192.168.99.100:2376"
    export DOCKER_CERT_PATH="/Users/tingting/.docker/machine/machines/default"
    export DOCKER_MACHINE_NAME="default"
    # Run this command to configure your shell: 
    # eval "$(docker-machine env default)"

The second one, doesn't look good

    Incorrect Usage.
    
    Usage: docker-machine ssh [arg...]
    
    Log into or run a command on a machine with SSH.
    
    Description:
       Arguments are [machine-name] [command]

**Comment by FrenchBen** [comment]
It's the correct usage. Did you type it or copy and pasted?

**Comment by sf-wind** [comment]
I copy and pasted

**Comment by sf-wind** [comment]
Maybe this is the reason?

**Comment by FrenchBen** [comment]
What about `docker-machine ssh default` ?
I forgot to ask, what operating system are you on?

**Comment by sf-wind** [comment]
I can do:

    docker-machine ssh default

It shows a picture and then 

    Boot2Docker version 1.8.3, build master : af8b089 - Mon Oct 12 18:56:54 UTC 2015
    Docker version 1.8.3, build f4bf5c7
    docker@default:~$ 

Then I can type 

    docker ps -a

    CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

I use OSX

**Comment by FrenchBen** [comment]
@sf-wind What about:
`eval "$(docker-machine env default)"`
`docker ps -a`

**Comment by sf-wind** [comment]
Which terminal shall I do this?
Do it in the docker terminal result to this:

    $ eval "$(docker-machine env default)"
    $ docker ps -a
    The server probably has client authentication (--tlsverify) enabled. Please check your TLS client certification settings: Get https://192.168.99.100:2376/v1.20/containers/json?all=1: remote error: bad certificate

**Comment by FrenchBen** [comment]
the docker terminal should do this for you - I would suggest doing it in a regular terminal to test things out.

**Comment by sf-wind** [comment]
Yes, I did it in the docker terminal. I also tried it in regular terminal. They both show the same error message.

**Comment by FrenchBen** [comment]
after this command: `eval "$(docker-machine env default)"` can you type: `env | grep -i 'docker'`
It seems that everything should be working properly with the proper env

**Comment by sf-wind** [comment]
$ eval "$(docker-machine env default)"
    $ env | grep -i 'docker'
    DOCKER_HOST=tcp://192.168.99.100:2376
    DOCKER_MACHINE_NAME=default
    DOCKER_TLS_VERIFY=1
    DOCKER_CERT_PATH=/Users/tingting/.docker/machine/machines/default

I just cannot use docker command.

**Comment by FrenchBen** [comment]
This is very strange. Can you check hopefully one more thing for me?
`docker --version`
`docker-machine -version`

**Comment by sf-wind** [comment]
Here you are:

    $ docker --version
    Docker version 1.8.3, build f4bf5c7
    $ docker-machine -version
    docker-machine version 0.4.1 (e2c88d6)

**Comment by FrenchBen** [comment]
Can you install the latest toolbox and see if it helps?
https://www.docker.com/docker-toolbox

**Comment by sf-wind** [comment]
No, it doesn't help. 

What is the source of this issue?

**Comment by ivstas** [comment]
Hello. I also encounter the same problem. I just followed the docs: http://docs.docker.com/mac/step_one/

I use Mac Os X El Capitan. Docker-toolbox 1.8.3, docker-machine 0.4.1

**Comment by ivstas** [comment]
Btw, I recreated VM using `docker-machine` using default settings, and then inspected machine using `docker-machine` inspect. It appears that in `HostOptions.EngineOptions` `TlsVerify` is set to **true**, while `TlsCaCert`, `TlsCert` and `TlsKey` are empty (`""`).
Is it the correct behaviour?

**Comment by ivstas** [comment]
The problem remains on toolbox version 1.9.0d.
Docker log on default VM (/var/lib/boot2docker/docker.log) says: 
```
http: TLS handshake error from 192.168.99.1:54213: tls: failed to verify client's certificate: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate
```

**Comment by rustinpc** [comment]
I was having the same problem, but the only way I was able to solve it was to uninstall docker and when I installed it again using Homebrew it worked fine (I had followed the docs at http://docs.docker.com/mac/step_one/ the first time).

**Comment by ravishtiwari** [comment]
@FrenchBen - I've followed the instructions and suggestions given in this thread so far.
Here is information of my env : 
```
docker --version
Docker version 1.9.0, build 76d6bc9
```
and 
```
docker-machine -version
docker-machine version 0.5.0 (04cfa58)
```
I can do **docker-machine ssh default** without any issue, and it log me into docker shell. 
My **docker-machine env default**
```
export DOCKER_TLS_VERIFY="1"
export DOCKER_HOST="tcp://192.168.99.100:2376"
export DOCKER_CERT_PATH="/Users/ravishtiwari/.docker/machine/machines/default"
export DOCKER_MACHINE_NAME="default"
# Run this command to configure your shell: 
# eval "$(docker-machine env default)"
```
However as soon as I do 
```
docker run hello-world
The server probably has client authentication (--tlsverify) enabled. Please check your TLS client certification settings: Post https://192.168.99.100:2376/v1.21/containers/create: remote error: bad certificate
```
I've already regenerated the certs and followed other instructions from thread, I even reinstalled Toolbox, and I am getting:
```
EPROTO
EPROTO
write
```
When trying to add a container through **Kitematic**.
I am not sure why I am facing so many issues, I've used docker before on Ubuntu/Elementary OS, and I can't remember if I faced any issue.

**Comment by liuzheng712** [comment]
After you run  ` eval "$(docker-machine env default)"`

please run this following command

    export DOCKER_OPTS="-H $DOCKER_HOST --tls --tlskey $DOCKER_CERT_PATH/server-key.pem    --tlscert $DOCKER_CERT_PATH/server.pem --tlsveri|   ~                          
    fy --tlscacert $DOCKER_CERT_PATH/ca.pem "
    alias docker="docker $DOCKER_OPTS " 

Then the `docker` works fine, but I don't know why not append this to `docker-machine env default`

this works fine for me

**Comment by ravishtiwari** [comment]
@liuzheng712 - Thanks, this seems to have resolved my issue, but, I am still getting same error while try to create container through **Kitematic**.

**Comment by FrenchBen** [comment]
Sounds like your VM wasn't created properly - I would go ahead and close kitematic, remove it and then open Kitematic to re-create it.
To remove the VM:
`docker-machine stop default && docker-machine rm default`

**Comment by ravishtiwari** [comment]
@FrenchBen - I removed the VM, as suggested after stopping the **Kitematic**, as suggested. restarted **Kitematic**, tried creating the 'hello-world' container from there, still getting 
```
EPROTO
EPROTO
write
```
Looks like some permission issue to me, but, not sure which directory.

**Comment by FrenchBen** [comment]
@ravishtiwari Can you try the following:
`docker $(docker-machine config default) ps`
then:
`docker $(docker-machine config default) run hello-world`

**Comment by ravishtiwari** [comment]
@FrenchBen both command are giving me error:
```
docker $(docker-machine config default) ps
Please specify only one -H
```

**Comment by FrenchBen** [comment]
@ravishtiwari what about:
`docker-machine ssh default`

**Comment by ravishtiwari** [comment]
@FrenchBen - as strange as it could be, I can ssh to default :)
```
docker-machine ssh default
                        ##         .
                  ## ## ##        ==
               ## ## ## ## ##    ===
           /"""""""""""""""""\___/ ===
      ~~~ {~~ ~~~~ ~~~ ~~~~ ~~~ ~ /  ===- ~~~
           \______ o           __/
             \    \         __/
              \____\_______/
 _                 _   ____     _            _
| |__   ___   ___ | |_|___ \ __| | ___   ___| | _____ _ __
| '_ \ / _ \ / _ \| __| __) / _` |/ _ \ / __| |/ / _ \ '__|
| |_) | (_) | (_) | |_ / __/ (_| | (_) | (__|   <  __/ |
|_.__/ \___/ \___/ \__|_____\__,_|\___/ \___|_|\_\___|_|
Boot2Docker version 1.9.0, build master : 16e4a2a - Tue Nov  3 19:49:22 UTC 2015
Docker version 1.9.0, build 76d6bc9
docker@default:~$ 

```

**Comment by FrenchBen** [comment]
What's the output of: `docker-machine config default` and `docker-machine -version` and lastly `which docker-machine`

**Comment by ravishtiwari** [comment]
@FrenchBen - output
```commonapp$ docker-machine config default
--tlsverify --tlscacert="/Users/ravishtiwari/.docker/machine/certs/ca.pem" --tlscert="/Users/ravishtiwari/.docker/machine/certs/cert.pem" --tlskey="/Users/ravishtiwari/.docker/machine/certs/key.pem" -H=tcp://192.168.99.100:2376LAP-478
```
and 
```
docker-machine -version
docker-machine version 0.5.0 (04cfa58)
```
and finally
```
which docker-machine
/usr/local/bin/docker-machine
```

PS: no problem, but, I am having a feeling that Mac is not the platform I should have tried this on, it worked for me earlier on Ubuntu, and I should have stick to it.

**Comment by ivstas** [comment]
@liuzheng712 Thanks, also worked for me.

**Comment by liuzheng712** [comment]
@ravishtiwari I like terminal work , the GUI not my taste , but the principle should be same, the bash environment need add some variable. You can search the <b>Kitematic</b> source code and find the environment setting part, add the lines I give after  

    export DOCKER_TLS_VERIFY="1"
    export DOCKER_HOST="tcp://192.168.99.100:2376"
    export DOCKER_CERT_PATH="xxxxxx"
    export DOCKER_MACHINE_NAME="default"

maybe it will work.

If can't, you @me later. I will try to see the source code.

**Comment by FrenchBen** [comment]
@liuzheng712 thanks for chiming in - Running `eval "$(docker-machine env default)"` does exactly what you've setup above. 
It seems that for @ravishtiwari something is getting in the way of connectivity, preventing the connection.

ideally this should work without any hiccups:
```
$ eval "$(docker-machine env default)"
$ docker ps -a
$ docker run hello-world
```

**Comment by ravishtiwari** [comment]
@FrenchBen 
I think @ravishtiwari found the way to knock it out. perhaps **eval "$(docker-machine env default)"** works for me, but, terminal show no output, but, I was seeing the same --tslverify error on these commands
```
docker ps -a
docker run hello-world
```
So, I did this:

```
export DOCKER_OPTS="-H $DOCKER_HOST --tls --tlskey $DOCKER_CERT_PATH/server-key.pem    --tlscert $DOCKER_CERT_PATH/server.pem --tlsverify --tlscacert $DOCKER_CERT_PATH/ca.pem "
alias docker="docker $DOCKER_OPTS " 
docker ps -a
```
now docker correctly lists out the available containers, and I am able to run containers, such as 
```
docker run -it ubuntu bash
```
So, bottom is : if **export DOCKER_OPTS="-H $DOCKER_HOST --tls --tlskey $DOCKER_CERT_PATH/server-key.pem    --tlscert $DOCKER_CERT_PATH/server.pem --tlsverify --tlscacert $DOCKER_CERT_PATH/ca.pem "
alias docker="docker $DOCKER_OPTS " ** is so necessary, why not add this to **.profile** or **.bashrc** during docker setup?

I am going to add this to my **.profile** for now and see if that works for me. BTW, I still getting same error trying to add container from **Kitematic**, not sure why, but, it seems I am no longer going to use it, as I am not sure it's worth spending that much time. I've too much trying to figure out how to make it work, terminal works for me, and I am happy as long as it's working for me.

**Comment by FrenchBen** [comment]
@ravishtiwari Happy to hear you got this working. 
The export of DOCKER_OPTS that you're referring to should be the same as the eval. 
One thing to note is that you will not be able to use your export until you actually use the eval, since `$DOCKER_HOST` will be unknown.

The eval should be enough to get things working. Additionally most will click the 'Docker CLI' button in kitematic, which sets a bunch of other stuff too - What you're doing is using your own terminal/setup to interact with Docker, which is fine but does require some of the additional steps you've listed above.
Adding them to your bashrc/profile IMHO should be left to the user, as we do the proper config from within kitematic and depending on the user, they may have a very specific shell setup.

You can also us the 'Docker quistart terminal' app to get things setup, which is our version of the profile/bashrc, without being intrusive.

Since this has been resolved, I'm going to close this issue, but feel free, to re-open it if needed.

**Comment by liuzheng712** [comment]
@ravishtiwari 

So, bottom is : if export DOCKER_OPTS="-H $DOCKER_HOST --tls --tlskey $DOCKER_CERT_PATH/server-key.pem --tlscert $DOCKER_CERT_PATH/server.pem --tlsverify --tlscacert $DOCKER_CERT_PATH/ca.pem "
alias docker="docker $DOCKER_OPTS " is so necessary, why not add this to .profile or .bashrc during docker setup?

this totally WRONG !!!!!!!!

if you set it into .bashrc, the value in DOCKER_OPTS will get <b>NOTHING!!!!!</b> or <b>THE OLD VALUE</b>

because when you source it , the `eval "$(docker-machine env default)"` didn't run yet, that means the value ` $DOCKER_HOST ` , ` $DOCKER_CERT_PATH ` get <b>NOTHING</b> get <b>NOTHING</b> get <b>NOTHING</b>

**Comment by ivstas** [comment]
@liuzheng712 what about adding
```bash
eval "$(docker-machine env default)"
export DOCKER_OPTS="-H $DOCKER_HOST --tls --tlskey $DOCKER_CERT_PATH/server-key.pem --tlscert $DOCKER_CERT_PATH/server.pem --tlsverify --tlscacert $DOCKER_CERT_PATH/ca.pem"
alias docker="docker $DOCKER_OPTS"
```

This should work, shouldn't it?

@FrenchBen according to docker docs, adding new VM should be as easy as:
```bash
docker-machine create <name> --driver virtualbox
eval "$(docker-machine env <name>)"
```
and switching between machines should also be done via single command: `eval "$(docker-machine env <name>)"`

This should work **out-of-the-box**. I believe that if this does not work **out-of-the-box**, it's to early to close the issue.

**Comment by liuzheng712** [comment]
@ivstas ye~
We can add this to ~/.bashrc 

``` bash
function SomeFunctionNameYouLike(){
    eval "$(docker-machine env $1)"
    export DOCKER_OPTS="-H $DOCKER_HOST --tls --tlskey $DOCKER_CERT_PATH/server-key.pem --tlscert $DOCKER_CERT_PATH/server.pem --tlsverify --tlscacert $DOCKER_CERT_PATH/ca.pem"
    alias dockers="docker $DOCKER_OPTS"
}
```
if we do this `SomeFunctionNameYouLike default` it will run the following lines , and in case run many times that change the `docker` command too much , I suggest use `dockers` to replace `docker`.

**Comment by FrenchBen** [comment]
@ivstas I agree that it should work out of the box and I've never had any issues with it. 
I'm closing this as the OP did get it working and this is a machine issue, not a kitematic issue. Feel free to open the issue within https://github.com/docker/machine

@liuzheng712 Correct a lot of things to be added to .`bashrc` for it to work as expected. I appreciate the work-around provided which allowed for it to resolve the issue posted here.

**Comment by albertma** [comment]
export DOCKER_OPTS="-H $DOCKER_HOST --tls --tlskey $DOCKER_CERT_PATH/server-key.pem    --tlscert $DOCKER_CERT_PATH/server.pem --tlsverify --tlscacert $DOCKER_CERT_PATH/ca.pem "
alias docker="docker $DOCKER_OPTS "


This is works for me!

**Comment by FrenchBen** [comment]
@albertma this is what the `docker-machine env default` does for you.

**Comment by yidinghan** [comment]
@liuzheng712 hi, the export DOCKER_OPTS solve my problem, and thx.

```shell
 function SomeFunctionNameYouLike(){
    eval "$(docker-machine env $1)"
    export DOCKER_OPTS="-H $DOCKER_HOST --tls --tlskey $DOCKER_CERT_PATH/server-key.pem --tlscert $DOCKER_CERT_PATH/server.pem --tlsverify --tlscacert $DOCKER_CERT_PATH/ca.pem"
    alias dockers="docker $DOCKER_OPTS"
}
```

there is another issue i think u must got an wonderful solution.
what i have to do when i want to use docker-compose but also meet the same issue with ssl error?


```shell
docker-compose up
ERROR: SSL error: [SSL: SSLV3_ALERT_BAD_CERTIFICATE] sslv3 alert bad certificate (_ssl.c:581)
```

i have already try to use  DOCKER_OPTS to solve the issue but doesn't work.

```shell
function dm(){
    eval "$(docker-machine env $1)"
    export DOCKER_OPTS="-H $DOCKER_HOST --tls --tlskey $DOCKER_CERT_PATH/server-key.pem --tlscert $DOCKER_CERT_PATH/server.pem --tlsverify --tlscacert $DOCKER_CERT_PATH/ca.pem"
    alias ds="docker $DOCKER_OPTS"
    alias dcs="docker-compose $DOCKER_OPTS"
}
```

**Comment by liuzheng712** [comment]
@yidinghan sorry , I'm never use this command before.
Recently I'm busy in doing my work, later I'll find time to lean it.
I need to know how `docker-compose` get environment variables, then find the solution.

**Comment by yidinghan** [comment]
@liuzheng712 thanks for reply.
It is so thankful if your could spend ur time to help me, but is there any document or file, that could help me to find out by myself, I may need to look at first.

**Comment by liuzheng712** [comment]
@yidinghan you need try to read the code, and understand how it works
and you're Chinese , so i recommend you this book ISBN:978711510727 《docker 源码分析》--孙宏亮
I don't know if  there have the answer you want, I didn't finish it

**Comment by yidinghan** [comment]
@liuzheng712 Ok, wish you could, in some day, looks as cute as flowers and like me cute also.（误

**Comment by nyarly** [comment]
I think my causes were unrelated to the above, but I got very similar issues post regenerating certificates - I believe but am not sure that this was the result of creating my own certificates in an effort to get a registry working within docker-machine. The fix for me was to `docker-machine rm` the machine I was working on, then `rm ~/.docker/machine/certs/*`, and finally `docker-machine create ...`

**Comment by FrenchBen** [comment]
If you are over-writing the certs it can definitely cause some issues.

**Comment by marcellodesales** [comment]
The `DOCKER_OPTS ` described above worked for me!!! Managing Docker Engines created manually is  a needed pain.

**Comment by rmanasek** [comment]
My issue was that I installed minikube which created cluster with IP 192.168.99.100 (and docker machine default was offline at that time). After I started docker machine "default" it got assigned 192.168.99.101. which could mean that generated certificated were not valid anymore for this new IP. I fixed it this way: 
- stopped minikube cluster
- stopped docker machine
- reboot my windows OS
- started docker machine - check that assigned IP is 192.168.99.100
- optionally: docker-machine -D regenerate-certs default
Good luck!
